### PR TITLE
Add python unitest for op

### PR DESCRIPTION
**Pull request by zhangting2020**
添加单测基类OpTest，所有的Op测试都继承该类。使用方式：
```python
class TestReluOp(OpTest):
```
- 需要实现一个TestxxOp的class，主要涉及的函数说明：
  -  `setUp`：每个test执行前会调用，基本的实现可以参考PR中的单测案例，如果有其他需要可以自己扩展。
  ```
    def setUp(self):
        self.init_case()
  ```
  - `init_case`：主要是对feed的数据、OP的参数配置做初始化。不同的单测case，只需要写不同的init_case
  ```
    def init_case(self):
        self.inputs = {
            "x": np.random.random([
                32,
                64,
            ]).astype("float32"),
            "dout": np.ones((32, 64)).astype("float32")
        }
  ```
  - `build_paddle_program`：实现paddle的前向，并调用基类的get_paddle_grads方法计算反向
  ```
    def build_paddle_program(self, target):
        x = paddle.to_tensor(self.inputs["x"], stop_gradient=False)
        out = F.relu(x)

        self.paddle_outputs = [out]
        self.paddle_grads = self.get_paddle_grads([out], [x],
                                                  [self.inputs["dout"]])
  ```
  - `build_cinn_program`：实现cinn的前向+反向，并调用基类的get_cinn_output方法，分别拿到前、反向的输出。**本意是希望将前、反向算子写在同一个program中，但测试发现relu融合后，仅计算了反向输出，因此relu暂时分成2个program。add的前、反向在同一个program没有问题。**
  ```
    def build_cinn_program(self, target):
        builder = NetBuilder("relu")
        x = builder.create_input(Float(32), self.inputs["x"].shape, "x")
        out = builder.relu(x)
        prog = builder.build()
        forward_res = self.get_cinn_output(prog, target, [x],
                                           [self.inputs["x"]], [out])

        builder = NetBuilder("relu_grad")
        shape = self.inputs["dout"].shape
        dout = builder.create_input(Float(32), shape, "dout")
        out = builder.create_input(Float(32), shape, "out")
        x_grad = builder.relu_grad(dout, out)
        prog = builder.build()
        backward_res = self.get_cinn_output(
            prog, target, [dout, out], [self.inputs["dout"], forward_res[0]],
            [x_grad])

        self.cinn_outputs = forward_res
        self.cinn_grads = backward_res
  ```

**更加复杂的情况可以参考test_elementwise_add_op.py，其中涉及了cinn和paddle算子的对齐，以及多个case如何写test**

**skip_if**

- 可以使用装饰器，跳过单测执行。
```
@OpTestTool.skip_if(not is_compiled_with_cuda(),
                    "x86 test will be skipped due to timeout.")
```
- 因目前发现新增的3个test会在x86 CI上timeout，因此暂时禁用，log如下
```
/usr/lib/python3/dist-packages/setuptools/depends.py:2: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
s
----------------------------------------------------------------------
Ran 1 test in 0.000s

OK (skipped=1)
```

### CODE REVIEW

File: python/CMakeLists.txt [hunk 1]
```diff
@@ -114,5 +114,21 @@ ADD_TEST(NAME test_cinn_real_squeezenet
 ADD_TEST(NAME test_cinn_real_facedet
     COMMAND ${CMAKE_COMMAND} -E env PYTHONPATH=${CMAKE_BINARY_DIR}/python:$ENV{PYTHONPATH}
     python3 ${CMAKE_CURRENT_SOURCE_DIR}/tests/test_facedet.py "${CMAKE_BINARY_DIR}/thirds/FaceDet" "${WITH_CUDA}"
+
+ADD_TEST(NAME test_relu_op
```
Comment (Xreki) [inline python/CMakeLists.txt]: 算子的单测，要不要新建一个子目录？另外，可以封装一个function简化一下cmake。

File: python/tests/op_test.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            if len(self.cinn_grads) != 0:
+                print("============ Check Grads ============")
+                self.check_results(self.paddle_grads, self.cinn_grads)
+
+    def check_results(self, expect_res, actual_res):
+        self.assertEqual(len(expect_res), len(actual_res))
+        for i in range(len(expect_res)):
+            print("Check the %d -th Result..." % i)
+            self.assertTrue(
+                np.allclose(expect_res[i], actual_res[i], atol=1e-4))
```
Comment (Xreki) [inline python/tests/op_test.py]: float要求至少1e-5。

File: python/tests/test_elementwise_add_op.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    def build_cinn_program(self, target):
+        builder = NetBuilder("add")
+        x = builder.create_input(Float(32), self.inputs["x"].shape, "x")
+        y = builder.create_input(Float(32), self.inputs["y"].shape, "y")
+        out = builder.elementwise_add(x, y, axis=self.axis)
+        prog = builder.build()
+        forward_res = self.get_cinn_output(
+            prog, target, [x, y], [self.inputs["x"], self.inputs["y"]], [out])
+
+        builder = NetBuilder("add_grad")
```
Comment (Xreki) [inline python/tests/test_elementwise_add_op.py]: 加个comment，这里为啥要用2个NetBuilder？

File: python/CMakeLists.txt (hunk 1 above)
Comment (zhangting2020) [reply python/CMakeLists.txt]: done. 已经单独放到python/test/ops下了。cmake已简化

File: python/tests/op_test.py (hunk 2 above)
Comment (zhangting2020) [reply python/tests/op_test.py]: 改为1e-6了

File: python/tests/test_elementwise_add_op.py (hunk 3 above)
Comment (zhangting2020) [reply python/tests/test_elementwise_add_op.py]: - 在relu的单测中加了说明
- add将前反向写到同一个program测试无问题，这个文件已改为一个NetBuilder

File: python/tests/ops/test_elementwise_add_op.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    def setUp(self):
+        self.init_results()
+        self.init_case()
+        self.init_target()
+
+    def init_case(self):
+        self.inputs = {
+            "x": np.random.random([32, 64]).astype("float32"),
+            "y": np.random.random([32, 64]).astype("float32"),
+            "dout": np.ones((32, 64)).astype("float32")
```
Comment (Xreki) [inline python/tests/ops/test_elementwise_add_op.py]: dout也可以设置成随机数？

File: python/tests/ops/test_relu_op.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import cinn
+from cinn.frontend import *
+from cinn.common import *
+
+
+class TestReluOp(OpTest):
+    def setUp(self):
+        self.init_results()
+        self.init_case()
+        self.init_target()
```
Comment (Xreki) [inline python/tests/ops/test_relu_op.py]: `init_targets`和`init_results`可以不需在每个op单测中显式调用？

File: python/tests/test_relu_op.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
```
Comment (Xreki) [inline python/tests/test_relu_op.py]: 这个文件忘了删除了。

File: python/tests/test_relu_op.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    def build_paddle_program(self, target):
+        x = paddle.to_tensor(self.inputs["x"], stop_gradient=False)
+        out = F.relu(x)
+
+        self.paddle_outputs = [out]
+        self.paddle_grads = self.get_paddle_grads([out], [x],
+                                                  [self.inputs["dout"]])
+
+    # Note: If the forward and backward operators are run in the same program,
+    # the forward result will be incorrect.
```
Comment (Xreki) [inline python/tests/test_relu_op.py]: 注释应该加到ops目录下的test_relu_op.py？

File: python/tests/ops/op_test.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            outs_and_grads.append(res.numpy(target))
+
+        return outs_and_grads
+
+    def apply_pass(self, prog, target):
+        print("============ Before Decomposer Pass ============")
+        for i in range(prog.size()):
+            print(prog[i])
+
+        prog.apply_pass(target, ["Decomposer"])
```
Comment (Superjomn) [inline python/tests/ops/op_test.py]: 可以加个argument，`passes=[]` ，便于外界设置

File: python/tests/ops/op_test.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        outs_and_grads = []
+        for res in result:
+            outs_and_grads.append(res.numpy(target))
+
+        return outs_and_grads
+
+    def apply_pass(self, prog, target):
+        print("============ Before Decomposer Pass ============")
+        for i in range(prog.size()):
+            print(prog[i])
```
Comment (Superjomn) [inline python/tests/ops/op_test.py]: `print` 改成 log吧，默认是不是不需要打印出来？
不然 CI log 也有点多

**Review by Superjomn — COMMENTED** [review]
Consider to add some general OpTester like `BinaryOpTester` and `UnaryOpTester` latter, for there are some routines between the `relu`, `prelu`, `sigmoid` Ops' definitions.

This PR is ok for the current situation.

File: python/tests/ops/op_test.py (hunk 9 above)
Comment (zhangting2020) [reply python/tests/ops/op_test.py]: done. 使用logging，仅在debug level下打印

File: python/tests/ops/op_test.py (hunk 8 above)
Comment (zhangting2020) [reply python/tests/ops/op_test.py]: - apply_pass增加默认参数passes=["Decomposer"]
- **TODO: general OpTester like BinaryOpTester and UnaryOpTester will be added in next PR.**

File: python/tests/test_relu_op.py (hunk 7 above)
Comment (zhangting2020) [reply python/tests/test_relu_op.py]: done.

File: python/tests/ops/test_elementwise_add_op.py (hunk 4 above)
Comment (zhangting2020) [reply python/tests/ops/test_elementwise_add_op.py]: done.

**Review by Superjomn — DISMISSED** [review]
LGTM

**Review by Xreki — DISMISSED** [review]
LGTM and good work~

**Review by Xreki — APPROVED** [review]
